File: android/src/main/java/com/tiktokeventsdk/HybridTiktokEventSdk.kt
package com.tiktokeventsdk

import com.margelo.nitro.core.Promise
import com.margelo.nitro.tiktokeventsdk.HybridTiktokEventSdkSpec
import com.tiktok.TikTokBusinessSdk.*;
import com.tiktok.TikTokBusinessSdk.TTConfig;
import com.tiktok.TikTokBusinessSdk.TTInitCallback
import com.tiktok.appevents.base.TTBaseEvent

class CallBack: TTInitCallback {
    override fun success() {
        TODO("Not yet implemented")
    }

    override fun fail(code: Int, msg: String?) {
        TODO("Not yet implemented")
    }

}

class HybridTiktokEventSdk: HybridTiktokEventSdkSpec() {
    override fun initialSdk(accessToken: String, appId: String, tiktokAppId: String): Promise<Boolean> {
        val prs = Promise<Boolean>()
        val config = TTConfig(getApplicationContext(), accessToken).setAppId(appId).setTTAppId(tiktokAppId).enableAutoIapTrack()
        initializeSdk(config, object: TTInitCallback {
            override fun success() {
                prs.resolve(true)
            }

            override fun fail(code: Int, msg: String?) {
                prs.reject(Throwable(msg))
            }

        })
        startTrack()
        return prs
    }

    override fun identifyUser(userId: String, userName: String?, phone: String?, email: String?) {
        identifyUser(userId, userName, phone, email)
    }

    override fun logout() {
        logout()
    }
    
    override fun trackEvent(event: String, params: Map<String, String>?) {
        if (params == null) return
        val eventBuilder = TTBaseEvent.newBuilder(event)
        params.keys.forEach {
            eventBuilder.addProperty(it, params[it])
        }
        trackTTEvent(eventBuilder.build())
    }
}

File: ios/HybridTiktokEventSdk.swift
//
//  HybridTiktokEventSdk.swift
//  Pods
//
//  Created by bachvq on 9/29/2025.
//

import Foundation
import TikTokBusinessSDK
import NitroModules

class HybridTiktokEventSdk: HybridTiktokEventSdkSpec {
    func initialSdk(accessToken: String, appId: String, tiktokAppId: String) throws -> Promise<Bool> {
        let prs = Promise<Bool>.init()
        TikTokBusiness.initializeSdk(.init(accessToken: accessToken, appId: appId, tiktokAppId: tiktokAppId)) { success, err in
            if let err {
                print("TIKTOK INITIAL FAILED", err)
                prs.reject(withError: err)
            } else {
                prs.resolve(withResult: success)
            }
        }
        return prs
    }
    
    func identifyUser(userId: String, userName: String?, phone: String?, email: String?) throws {
        TikTokBusiness.identify(withExternalID: userId, externalUserName: userName, phoneNumber: phone, email: email)
    }
    
    func logout() throws {
        TikTokBusiness.logout()
    }
    
    func trackEvent(event: String, params: Dictionary<String, String>?) throws {
        TikTokBusiness.trackTTEvent(.init(eventName: event, properties: params ?? .init(), eventId: nil))
    }
}
